File: bindings/nodejs/lib/locality.js
/*
 * locality.js
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


"use strict";

var buffer = require('./bufferConversion');
var transactional = require('./retryDecorator');
var Database = require('./database');
var LazyIterator = require('./lazyIterator');
var fdb = require('./fdbModule');
var fdbUtil = require('./fdbUtil');

var KEY_SERVERS_PREFIX = buffer.fromByteLiteral('\xff/keyServers/');
var PAST_VERSION_ERROR_CODE = 1007;

function getBoundaryKeysImpl(tr, begin, end, callback) {
	function BoundaryFetcher(wantAll) {
		this.tr = tr;
		this.begin = begin;
		this.end = end;
		this.lastBegin = undefined;

		var fetcher = this;

		if(wantAll)
			this.streamingMode = fdb.streamingMode.wantAll;
		else
			this.streamingMode = fdb.streamingMode.iterator;

		function iteratorCb(err, res) {
			if(err) {
				if(err.code === PAST_VERSION_ERROR_CODE && fetcher.begin !== fetcher.lastBegin) {
					fetcher.tr = fetcher.tr.db.createTransaction();
					readKeys();
				}
				else {
					fetcher.tr.onError(err, function(e) {
						if(e)
							fetcher.fetchCb(e);
						else
							readKeys();
					});
				}
			}
			else
				fetcher.fetchCb();
		}

		function readKeys() {
			fetcher.lastBegin = fetcher.begin;
			fetcher.tr.options.setReadSystemKeys();
			fetcher.tr.options.setLockAware();
			fetcher.tr.snapshot.getRange(fetcher.begin, fetcher.end, {streamingMode: fetcher.streamingMode}).forEachBatch(function(kvs, innerCb) {
				fetcher.forEachCb = innerCb;
				var keys = kvs.map(function(kv) { return kv.key.slice(13); });
				var last = kvs[kvs.length-1].key;
				fetcher.begin = Buffer.concat([last, buffer.fromByteLiteral('\x00')], last.length + 1);
				fetcher.fetchCb(undefined, keys);
			}, iteratorCb);

			fetcher.streamingMode = fdb.streamingMode.wantAll;
		}

		this.fetch = function(cb) {
			this.fetchCb = cb;
			if(this.read)
				this.forEachCb();
			else {
				this.read = true;
				readKeys();
			}
		};

		this.clone = function(wantAll) {
			var clone = new BoundaryFetcher(wantAll);

			clone.tr = this.tr.db.createTransaction();
			clone.begin = this.begin;
			clone.end = this.end;
			clone.lastBegin = this.lastBegin;

			return clone;
		};
	}

	callback(null, new LazyIterator(BoundaryFetcher));
}

function getBoundaryKeys(databaseOrTransaction, begin, end, callback) {
	begin = fdbUtil.keyToBuffer(begin);
	end = fdbUtil.keyToBuffer(end);

	begin = Buffer.concat([KEY_SERVERS_PREFIX, begin], KEY_SERVERS_PREFIX.length + begin.length);
	end = Buffer.concat([KEY_SERVERS_PREFIX, end], KEY_SERVERS_PREFIX.length + end.length);

	if(databaseOrTransaction instanceof Database) {
		getBoundaryKeysImpl(databaseOrTransaction.createTransaction(), begin, end, callback);
	}
	else {
		var tr = databaseOrTransaction.db.createTransaction();
		databaseOrTransaction.getReadVersion(function(err, ver) {
			tr.setReadVersion(ver);
			getBoundaryKeysImpl(tr, begin, end, callback);
		});
	}
}

var getAddressesForKey = transactional(function (tr, key, cb) {
	key = fdbUtil.keyToBuffer(key);
	tr.tr.getAddressesForKey(key, cb);
});

module.exports = {getBoundaryKeys: getBoundaryKeys, getAddressesForKey: getAddressesForKey};

File: bindings/python/fdb/locality.py
#
# locality.py
#
# This source file is part of the FoundationDB open source project
#
# Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# FoundationDB Python API

"""Documentation for this API can be found at
https://apple.github.io/foundationdb/api-python.html"""

from fdb import impl as _impl


def _get_boundary_keys(db_or_tr, begin, end):
    if isinstance(db_or_tr, _impl.Transaction):
        tr = db_or_tr.db.create_transaction()
        # This does not guarantee transactionality because of the exception handling below,
        # but it does hide latency for the new transaction's start
        tr.set_read_version(db_or_tr.get_read_version().wait())
    else:
        tr = db_or_tr.create_transaction()
    first_time = True
    while begin < end:
        try:
            lastbegin = begin
            tr.options.set_read_system_keys()
            tr.options.set_lock_aware()
            kvs = tr.snapshot.get_range(b'\xff' + b'/keyServers/' + begin, b'\xff' + b'/keyServers/' + end)
            if first_time:
                first_time = False
                yield None  # trick to get the above get_range to be asynchronously dispatched before get_boundary_keys() returns.
            for kv in kvs:
                yield kv.key[13:]
                begin = kv.key[13:] + b'\x00'
            begin = end
        except _impl.FDBError as e:
            # if we get a transaction_too_old and *something* has happened, then we are no longer transactional
            if e.code == 1007 and begin != lastbegin:
                tr = tr.db.create_transaction()
            else:
                tr.on_error(e).wait()


def get_boundary_keys(db_or_tr, begin, end):
    begin = _impl.keyToBytes(begin)
    end = _impl.keyToBytes(end)

    gen = _get_boundary_keys(db_or_tr, begin, end)
    try:
        next(gen)
    except StopIteration:  # if _get_boundary_keys() never yields a value, e.g. begin > end
        return (x for x in list())
    return gen


@_impl.transactional
def get_addresses_for_key(tr, key):
    keyBytes = _impl.keyToBytes(key)
    return _impl.FutureStringArray(tr.capi.fdb_transaction_get_addresses_for_key(tr.tpointer, keyBytes, len(keyBytes)))
